providers/atlassian.go

```diff
@@ -0,0 +1,239 @@
+	Picture         string            `json:"picture"`
+	AccountStatus   string            `json:"account_status"`
+	Nickname        string            `json:"nickname"`
+	Locale          string            `json:"locale"`
+	ExtendedProfile map[string]string `json:"extended_profile"`
+	EmailVerified   bool              `json:"email_verified"`
+}
+
+type resource struct {
+	ID        string   `json:"id"`
+	URL       string   `json:"url"`
+	Name      string   `json:"name"`
+	Scopes    []string `json:"scopes"`
+	AvatarURL string   `json:"avatarUrl"`
+}
+
+type resources []resource
+
+type group struct {
+	Name string `json:"name"`
+	Self string `json:"self"`
+}
+
+type groups []group
+
+const (
+	atlassianProviderName = "Atlassian"
+	atlassianDefaultScope = "read:jira-user read:me"
+	atlassianPrompt       = "consent"
+	atlassianAudience     = "api.atlassian.com"
+)
+
+var (
+	// Default Login URL for Atlassian.
+	// Pre-parsed URL of https://auth.atlassian.com/authorize.
+	atlassianDefaultLoginURL = &url.URL{
+		Scheme: "https",
+		Host:   "auth.atlassian.com",
+		Path:   "/authorize",
+	}
+
+	// Default Redeem URL for Atlassian.
+	// Pre-parsed URL of https://auth.atlassian.com/oauth/access_token.
+	atlassianDefaultRedeemURL = &url.URL{
+		Scheme: "https",
+		Host:   "auth.atlassian.com",
+		Path:   "/oauth/token",
+	}
+
+	// Default Validation URL for Atlassian.
+	// ValidationURL is the API Base URL.
+	// Other API requests are based on this.
+	// Pre-parsed URL of https://api.atlassian.com.
+	atlassianDefaultValidateURL = &url.URL{
+		Scheme: "https",
+		Host:   "api.atlassian.com",
+		Path:   "/",
+	}
+)
+
+// NewAtlassianProvider initiates a new AtlassianProvider
+func NewAtlassianProvider(p *ProviderData) *AtlassianProvider {
+	p.setProviderDefaults(providerDefaults{
+		name:        atlassianProviderName,
+		loginURL:    atlassianDefaultLoginURL,
+		redeemURL:   atlassianDefaultRedeemURL,
+		profileURL:  nil,
+		validateURL: atlassianDefaultValidateURL,
+		scope:       atlassianDefaultScope,
+	})
+	p.Prompt = atlassianPrompt
+	return &AtlassianProvider{ProviderData: p}
+}
+
+func makeAtlassianHeader(accessToken string) http.Header {
+	// extra headers required by the Atlassian API when making authenticated requests
+	extraHeaders := map[string]string{
+		acceptHeader: acceptApplicationJSON,
+	}
+	return makeAuthorizationHeader(tokenTypeBearer, accessToken, extraHeaders)
+}
+
+// SetGroups sets Atlassian Groups
+func (p *AtlassianProvider) SetGroups(groups []string) {
+	p.Groups = groups
+}
+
+// EnrichSessionState updates the Email and User after the initial Redeem
+func (p *AtlassianProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	err := p.validateUserGroups(ctx, s)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// ValidateSessionState validates the AccessToken
+func (p *AtlassianProvider) ValidateSession(ctx context.Context, s *sessions.SessionState) bool {
```

Comment:
Are you able to implement the `RefreshSessionIfNeeded` (refresh token refresh call to redeem endpoint)?

I'm working on refactoring `RefreshSession` here: https://github.com/oauth2-proxy/oauth2-proxy/pull/1086

I want to make sure our provider implementations stop skipping implementing that if possible. Users who use the `--cookie-refresh` option would be surprised when access tokens aren't refreshed like expected.

There's also a nasty performance bug at the moment when `ValidateSession` is implemented and `RefreshSessionIfNeeded` is not implemented and a `--cookie-refresh` is set that the validate endpoint will be hit on every request after the refresh timer is up.